### PR TITLE
ring: Use `net.JoinHostPort` to support IPv6 addresses

### LINKED ISSUE #3743: Add support for IPv6
#### Is your feature request related to a problem? Please describe.

IPv6 is a major blocker for organizations implementing the Grafana stack in single- or dual-stack IPv6 environments.

EKS now has full support for single-stack IPv6 clusters and we'll continue to see more and more IPv6 adoption.

#### Describe the solution you'd like

Mimir should be able to start with IPv6 enabled and:

- Either bind on [::0] or successfully identify and bind on a public IPv6 address attached to a network interface

- Members of the memberlist should be able to advertise their IPv6 address

- Memberlist should be able to properly communicate amongst itself using IPv6

- Components should query for both A and AAAA records and use AAAA records if IPv6 is enabled

#### Describe alternatives you've considered
N/A

#### Additional context

Support for this primarily depends on https://github.com/grafana/dskit/pull/185 being merged to support IPv6 in ring/memberlist

A working patch of Mimir with the augmented dskit and all necessary changes to support IPv6 is available here:
https://github.com/pauwels-labs/mimir/pull/1

If the Grafana team is interested in including this I'm happy to take ownership of the merging process to get this in ASAP once the dskit PR is merged.

### PR DESCRIPTION
**Pull request by matthewpi**
#### What this PR does

Fixes an issue when using IPv6 where IPv6 addresses are not properly joined with ports resulting in the ring attempting to incorrectly use `::1:7496` instead of the proper `[::1]:7496` notation.

With this change it's fully possible to use Mimir within an IPv6 only network.

#### Which issue(s) this PR fixes or relates to

Fixes #3743

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

---

For those interested in trying this out, here is a shortened snippet of the configuration changes I needed to make in order to get IPv6 working.

```yaml
alertmanager:
  sharding_ring:
    instance_enable_ipv6: true
compactor:
  sharding_ring:
    instance_enable_ipv6: true
distributor:
  ring:
    instance_enable_ipv6: true
ingester:
  ring:
    instance_enable_ipv6: true
memberlist:
  bind_addr:
    - '::'
  bind_port: 7946
  cluster_label: mimir
query_scheduler:
  service_discovery_mode: ring
  ring:
    instance_enable_ipv6: true
ruler:
  ring:
    instance_enable_ipv6: true
store_gateway:
  sharding_ring:
    instance_enable_ipv6: true
overrides_exporter:
  ring:
    instance_enable_ipv6: true
```

### CODE REVIEW

File: pkg/scheduler/schedulerdiscovery/ring_test.go [hunk 1]
```diff
@@ -50,7 +51,7 @@ func TestRingConfig_CustomConfigToBasicLifecyclerConfig(t *testing.T) {
 	// ring config
 	expected := ring.BasicLifecyclerConfig{
 		ID:                              "test",
-		Addr:                            "1.2.3.4:10",
+		Addr:                            net.JoinHostPort(cfg.InstanceAddr, strconv.Itoa(cfg.InstancePort)),
```
Comment (charleskorn) [inline pkg/scheduler/schedulerdiscovery/ring_test.go:54]: Is this change required?

File: pkg/scheduler/schedulerdiscovery/ring_test.go [hunk 2]
```diff
@@ -17,12 +18,12 @@ import (
...
 	cfg := RingConfig{}
 	flagext.DefaultValues(&cfg)
-	cfg.InstanceAddr = "127.0.0.1"
+	cfg.InstanceAddr = "::1"
 	cfg.InstancePort = 9095
 
 	expected := ring.BasicLifecyclerConfig{
 		ID:                              cfg.InstanceID,
-		Addr:                            fmt.Sprintf("%s:%d", cfg.InstanceAddr, cfg.InstancePort),
+		Addr:                            net.JoinHostPort(cfg.InstanceAddr, strconv.Itoa(cfg.InstancePort)),
```
Comment (charleskorn) [inline pkg/scheduler/schedulerdiscovery/ring_test.go:26]: Rather than duplicating the logic between the production code and the test, should we specify the desired behaviour (eg. `Addr` is set to `"[::1]:9095"`) here?

**Review by charleskorn — COMMENTED** [review]
Thanks for the PR @matthewpi! Could you please add an entry to the changelog for this?

File: pkg/scheduler/schedulerdiscovery/ring_test.go (hunk 1 above)
Comment (matthewpi) [reply pkg/scheduler/schedulerdiscovery/ring_test.go]: I reverted this change but added a separate test, that independently tests just IPv4 and IPv6.

**Comment by matthewpi** [comment]
> Thanks for the PR @matthewpi! Could you please add an entry to the changelog for this?

Do you have any suggestions for what the changelog entry should be?  Both in terms of what type (`CHANGE`, `BUGFIX`, etc) and the description.  I feel the title of this PR is useful for commit history, but maybe not for the changelog when people are wondering if this means IPv6 is supported.

**Comment by 56quarters** [comment]
> Do you have any suggestions for what the changelog entry should be? Both in terms of what type (`CHANGE`, `BUGFIX`, etc) and the description. I feel the title of this PR is useful for commit history, but maybe not for the changelog when people are wondering if this means IPv6 is supported.

What do you think about:

```
* [BUGFIX] Ring: Ensure network addresses used for component hash rings are formatted correctly when using IPv6. #6068
```

**Review by 56quarters — APPROVED** [review]
Looks great, thanks!
